File: api/src/test/java/net/blueshell/api/validation/UserValidatorTest.java
package net.blueshell.api.validation;

import jakarta.validation.ConstraintViolation;
import jakarta.validation.Validator;
import net.blueshell.api.dto.user.AdvancedUserDTO;
import net.blueshell.api.dto.user.SimpleUserDTO;
import net.blueshell.api.factory.dto.user.AdvancedUserDTOFactory;
import net.blueshell.api.factory.dto.user.SimpleUserDTOFactory;
import net.blueshell.api.validation.group.Creation;
import net.blueshell.api.validation.group.Update;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * Unit tests for user DTO validation groups.
 */
@SpringBootTest
class UserValidatorTest {

    @Autowired private Validator validator;
    @Autowired private SimpleUserDTOFactory simpleUserFactory;
    @Autowired private AdvancedUserDTOFactory advancedUserFactory;

    @Test
    void validSimpleUserDTO_passesValidation() {
        SimpleUserDTO dto = simpleUserFactory.createBasic();
        Set<ConstraintViolation<SimpleUserDTO>> violations = validator.validate(dto, Creation.class);
        assertTrue(violations.isEmpty(), "Valid SimpleUserDTO should pass validation");
    }

    @Test
    void simpleUserDTO_withoutPassword_failsCreationValidation() {
        SimpleUserDTO dto = simpleUserFactory.createWithCustomizations(u -> u.setPassword(null));
        Set<ConstraintViolation<SimpleUserDTO>> violations = validator.validate(dto, Creation.class);
        assertFalse(violations.isEmpty());
        assertTrue(violations.stream().anyMatch(v -> v.getPropertyPath().toString().equals("password")));
    }

    @Test
    void simpleUserDTO_withWeakPassword_failsValidation() {
        SimpleUserDTO dto = simpleUserFactory.createWithCustomizations(u -> u.setPassword("weak"));
        Set<ConstraintViolation<SimpleUserDTO>> violations = validator.validate(dto, Creation.class);
        assertFalse(violations.isEmpty());
        assertTrue(violations.stream().anyMatch(v -> v.getPropertyPath().toString().equals("password")));
    }

    @Test
    void simpleUserDTO_withInvalidEmail_failsValidation() {
        SimpleUserDTO dto = simpleUserFactory.createWithCustomizations(u -> u.setEmail("invalid-email"));
        Set<ConstraintViolation<SimpleUserDTO>> violations = validator.validate(dto);
        assertFalse(violations.isEmpty());
        assertTrue(violations.stream().anyMatch(v -> v.getPropertyPath().toString().equals("email")));
    }

    @Test
    void simpleUserDTO_withEmptyUsername_failsValidation() {
        SimpleUserDTO dto = simpleUserFactory.createWithCustomizations(u -> u.setUsername(""));
        Set<ConstraintViolation<SimpleUserDTO>> violations = validator.validate(dto);
        assertFalse(violations.isEmpty());
        assertTrue(violations.stream().anyMatch(v -> v.getPropertyPath().toString().equals("username")));
    }

    @Test
    void validAdvancedUserDTO_passesValidation() {
        AdvancedUserDTO dto = advancedUserFactory.createBasic();
        Set<ConstraintViolation<AdvancedUserDTO>> violations = validator.validate(dto, Update.class);
        assertTrue(violations.isEmpty(), "Valid AdvancedUserDTO should pass validation");
    }

    @Test
    void advancedUserDTO_withoutDateOfBirth_failsValidation() {
        AdvancedUserDTO dto = advancedUserFactory.createWithCustomizations(u -> u.setDateOfBirth(null));
        Set<ConstraintViolation<AdvancedUserDTO>> violations = validator.validate(dto);
        assertFalse(violations.isEmpty());
        assertTrue(violations.stream().anyMatch(v -> v.getPropertyPath().toString().equals("dateOfBirth")));
    }

    @Test
    void advancedUserDTO_withoutNationality_failsValidation() {
        AdvancedUserDTO dto = advancedUserFactory.createWithCustomizations(u -> u.setNationality(null));
        Set<ConstraintViolation<AdvancedUserDTO>> violations = validator.validate(dto);
        assertFalse(violations.isEmpty());
        assertTrue(violations.stream().anyMatch(v -> v.getPropertyPath().toString().equals("nationality")));
    }
}

File: api/src/test/kotlin/net/blueshell/api/user/validation/UserValidatorTest.kt
package net.blueshell.api.user.validation

import jakarta.validation.ConstraintViolation
import jakarta.validation.Validator
import net.blueshell.api.user.dto.AdvancedUserDTO
import net.blueshell.api.user.dto.SimpleUserDTO
import net.blueshell.api.factory.dto.user.AdvancedUserDTOFactory
import net.blueshell.api.factory.dto.user.SimpleUserDTOFactory
import net.blueshell.api.shared.validation.group.Creation
import net.blueshell.api.shared.validation.group.Update
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest

/**
 * Unit tests for user DTO validation groups.
 */
@SpringBootTest
class UserValidatorTest @Autowired constructor(
    private val validator: Validator,
    private val simpleUserFactory: SimpleUserDTOFactory,
    private val advancedUserFactory: AdvancedUserDTOFactory
) {

    @Test
    fun `valid simple user dto passes validation`() {
        val dto = simpleUserFactory.createBasic()
        val violations: Set<ConstraintViolation<SimpleUserDTO>> = validator.validate(dto, Creation::class.java)
        assertTrue(violations.isEmpty(), "Valid SimpleUserDTO should pass validation")
    }

    @Test
    fun `simple user dto without password fails creation validation`() {
        val dto = simpleUserFactory.createWithCustomizations { it.password = null }
        val violations: Set<ConstraintViolation<SimpleUserDTO>> = validator.validate(dto, Creation::class.java)
        assertFalse(violations.isEmpty())
        assertTrue(violations.any { it.propertyPath.toString() == "password" })
    }

    @Test
    fun `simple user dto with weak password fails validation`() {
        val dto = simpleUserFactory.createWithCustomizations { it.password = "weak" }
        val violations: Set<ConstraintViolation<SimpleUserDTO>> = validator.validate(dto, Creation::class.java)
        assertFalse(violations.isEmpty())
        assertTrue(violations.any { it.propertyPath.toString() == "password" })
    }

    @Test
    fun `simple user dto with invalid email fails validation`() {
        val dto = simpleUserFactory.createWithCustomizations { it.email = "invalid-email" }
        val violations: Set<ConstraintViolation<SimpleUserDTO>> = validator.validate(dto)
        assertFalse(violations.isEmpty())
        assertTrue(violations.any { it.propertyPath.toString() == "email" })
    }

    @Test
    fun `simple user dto with empty username fails validation`() {
        val dto = simpleUserFactory.createWithCustomizations { it.username = "" }
        val violations: Set<ConstraintViolation<SimpleUserDTO>> = validator.validate(dto)
        assertFalse(violations.isEmpty())
        assertTrue(violations.any { it.propertyPath.toString() == "username" })
    }

    @Test
    fun `valid advanced user dto passes validation`() {
        val dto = advancedUserFactory.createBasic()
        val violations: Set<ConstraintViolation<AdvancedUserDTO>> = validator.validate(dto, Update::class.java)
        assertTrue(violations.isEmpty(), "Valid AdvancedUserDTO should pass validation")
    }

    @Test
    fun `advanced user dto without date of birth fails validation`() {
        val dto = advancedUserFactory.createWithCustomizations { it.dateOfBirth = null }
        val violations: Set<ConstraintViolation<AdvancedUserDTO>> = validator.validate(dto)
        assertFalse(violations.isEmpty())
        assertTrue(violations.any { it.propertyPath.toString() == "dateOfBirth" })
    }

    @Test
    fun `advanced user dto without nationality fails validation`() {
        val dto = advancedUserFactory.createWithCustomizations { it.nationality = null }
        val violations: Set<ConstraintViolation<AdvancedUserDTO>> = validator.validate(dto)
        assertFalse(violations.isEmpty())
        assertTrue(violations.any { it.propertyPath.toString() == "nationality" })
    }
}
